### ISSUE TITLE
Query Information about specific GitHub Repo

### ISSUE THREAD

**Issue by HaRo87** [julien040/anyquery#16]
Hey @julien040,
first things first: I really love your project, kudos! 
I want to use your GitHub Plugin to query general information about a single repository which can be provided as an argument. I would be willing to contribute this if that would work for you. From what I understood from your contribution guidelines, this shouldn't be an issue. I would use plugins/github/repoListByUser.go as a reference for the implementation. I just wanted to check with you first before starting to contribute, if that would work for you.

Kept up the good work.

Cheers,
Robert

**Comment by julien040** [comment]
Hey Robert,
Thank you!

I’d love for your contribution to be added to the GitHub plugin. Feel free
to ask for help if you get stuck at any point; I’ll be happy to help you!

Best,
Julien

Le ven. 13 sept. 2024 à 09:05, Robert Hansel ***@***.***> a
écrit :

**Comment by HaRo87** [comment]
Hey @julien040,
thanks for offering your help. I have a first implementation and would like to know how I can test it locally. I managed to build my local version with `make all` and I can run `./main.out` but I think when I call `anyquery install [plugin]` I am not actually install the local version of the plugin but rather a remote (released) version. Is that correct? How do you test changes to the plugins?

Cheers,
Robert

**Comment by julien040** [comment]
Yes @HaRo87, when you run `anyquery install [plugin name]`, you're installing the released version.
I've written a [small guide](https://anyquery.dev/docs/developers/plugins/create-plugin/#debugging) on how to debug local plugins. But to sum up:

## 1. [Dev manifest](https://anyquery.dev/docs/developers/plugins/create-plugin/#the-manifest-file)

Create a [dev manifest](https://anyquery.dev/docs/developers/plugins/create-plugin/#the-manifest-file) in the plugin folder. `plugins/github/devManifest.json` in our case

```json
{
    "executable": "github.out",
    "build_command": "make",
    "user_config": {
        "default": {
            "token": "<your GitHub token>"
        }
    },
    "tables": [
        "my_repositories",
        "repositories_from_user",
        "commits_from_repository",
        "issues_from_repository",
        "pull_requests_from_repository",
        "releases_from_repository",
        "branches_from_repository",
        "contributors_from_repository",
        "tags_from_repository",
        "followers_from_user",
        "my_followers",
        "following_from_user",
        "my_following",
        "stars_from_user",
        "my_stars",
        "gists_from_user",
        "my_gists",
        "comments_from_issue",
        "my_issues",
        "stargazers_from_repository"
    ],
    "log_file": "dev.log",
    "log_level": "debug"
}
```

You'll need to add your new table in the array

## 2. Loading the dev manifest

Make sure you've uninstalled the GitHub plugin by running `anyquery profile delete default github default` and `anyquery plugin uninstall github`

Then, launch `anyquery` in development mode in the folder `plugins/github/`

```bash
anyquery --dev
```

And in the SQL shell, run

```sql
-- Run make and load the plugin
SELECT load_dev_plugin('github', 'devManifest.json');
```

All the tables listed in the dev manifest will be loaded.
When you do some changes to the plugin, you can reload it using

```sql
-- Unload the plugin, run make and load the plugin
SELECT reload_dev_plugin('github');
```

Congratulations, you've loaded a development plugin.

## 3. Adding your table

To add the repository info table that you want to create, add it in the array `tables` in the `devManifest.json` file
```js
   "tables": [
        "my_repositories",
        "repositories_from_user",
        "commits_from_repository",
        "issues_from_repository",
        "pull_requests_from_repository",
        ...
        "summary_of_repo" // Just a name example
]
```

Write your table in a new file. This guide explains what each function do :[https://anyquery.dev/docs/developers/plugins/create-plugin/#implement-a-table](https://anyquery.dev/docs/developers/plugins/create-plugin/#implement-a-table)

```go
// A constructor to create a new table instance
// This function is called everytime a new connection is made to the plugin
//
// It should return a new table instance, the database schema and if there is an error
func repoInfoCreator(args rpc.TableCreatorArgs) (rpc.Table, *rpc.DatabaseSchema, error) {
...
}
```

And in `main.go` of the plugin, add your function to `rpc.NewPlugin`

```go
package main

import (
	"github.com/julien040/anyquery/rpc"
)

func main() {
	plugin := rpc.NewPlugin(myRepoCreator, repoListByUserCreator, commitsCreator, issuesCreator, pullRequestsCreator,
		releaseCreator, branchesCreator, contributorsCreator, tagsCreator, followersCreator, myFollowerCreator,
		followingCreator, my_followingCreator, starsCreator, my_starsCreator, gistsCreator, my_gistsCreator, comments_from_issueCreator,
		my_issuesCreator, stargazers_repoCreator, /* Add here your new table */ repoInfoCreator)
	plugin.Serve()
}
```

You can find a long tutorial on plugins here: [https://anyquery.dev/docs/developers/plugins/create-plugin/](https://anyquery.dev/docs/developers/plugins/create-plugin/)

**Comment by julien040** [comment]
Hey @HaRo87, I saw your [commit](https://github.com/julien040/anyquery/commit/7036a8ed00599a9697b743fcddfd4dda6ed37c54) in the fork. 
Thank you for adding a new table. Do you want me to add the changes to the GitHub plugin ?

**Comment by HaRo87** [comment]
Hey @julien040
thanks for reaching out. I am not finished, yet. I am still facing some issues when testing but didn't find the time to continue working on it. I will try to come back to it this week and will open a PR once I am ready.
